### PR TITLE
Add operators for Deformable ConvNets/DFF

### PR DESCRIPTION
**Pull request by Orpine**

### CODE REVIEW

**Comment by piiswrong** [comment]
code looks in good shape.

two general comments: 
we don't indent in namespaces.
please add some tests for your operators.

**Comment by Orpine** [comment]
Hi @piiswrong , sorry for the delay. Now test code for psroipoling, deformable convolution/psroipooling has been added.

**Comment by piiswrong** [comment]
Could you also add a check_numerical_gradient test?

Its generally a good habit to check gradient. We have had people discovering their ops were outputing wrong gradient on corner cases when they submit PR. The model still trains with slightly wrong gradient but you often get better acc by correcting it.

**Comment by Orpine** [comment]
It seems that building check always reports `libdc1394 error: Failed to initialize libdc1394`

**Comment by futurely** [comment]
`sudo ln -s /dev/null /dev/raw1394`
https://github.com/dmlc/mxnet/blob/fd43d6880f0599d213dd3dd2d178a975fc9823b9/docker/tool.sh#L67

https://github.com/dmlc/mxnet/blob/70a0eeb14a508dec47460a0c289ca73eef27d6d7/docs/how_to/cloud.md#build-and-run-mxnet-on-a-gpu-instance

**Comment by sbodenstein** [comment]
@Orpine, @piiswrong: we would like to use this PR. What is still preventing it from being merged?

File: src/operator/contrib/deformable_convolution.cc [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+                                            std::vector<int> *in_type) const {
+  std::vector<TShape> out_shape, aux_shape;
+  std::vector<int> out_type, aux_type;
+  CHECK(InferType(in_type, &out_type, &aux_type));
+  CHECK(InferShape(in_shape, &out_shape, &aux_shape));
+  DO_BIND_DISPATCH(CreateOp, param_, (*in_type)[0], in_shape, &out_shape, ctx);
+}
+
+MXNET_REGISTER_OP_PROPERTY(_contrib_DeformableConvolution, DeformableConvolutionProp)
+.describe(R"code(Compute *N*-D convolution on *(N+2)*-D input.
```
Comment (piiswrong) [inline src/operator/contrib/deformable_convolution.cc]: These docs seems to be copied from standard conv. Could you updated it to describe DeformableConvolution or remove inaccurate doc and insert a link to the original paper?

File: src/operator/contrib/deformable_psroi_pooling.cc [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    std::vector<int> out_type, aux_type;
+    CHECK(InferType(in_type, &out_type, &aux_type));
+    CHECK(InferShape(in_shape, &out_shape, &aux_shape));
+    DO_BIND_DISPATCH(CreateOp, param_, in_type->at(0));
+  }
+
+  DMLC_REGISTER_PARAMETER(DeformablePSROIPoolingParam);
+
+  MXNET_REGISTER_OP_PROPERTY(_contrib_DeformablePSROIPooling, DeformablePSROIPoolingProp)
+    .describe("Performs region-of-interest pooling on inputs. Resize bounding box coordinates by "
```
Comment (piiswrong) [inline src/operator/contrib/deformable_psroi_pooling.cc]: same. updated/remove docs and link to paper

**Comment by piiswrong** [comment]
@sbodenstein @Orpine 
There are too many PRs I somehow overlooked this.

Will try to get it merged soon.

File: src/operator/contrib/nn/deformable_im2col.h [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+ * \param data_col column buffer pointer
+ */
+template <typename DType>
+inline void deformable_im2col(mshadow::Stream<cpu>* s,
+  const DType* data_im, const DType* data_offset,
+  const TShape& im_shape, const TShape& col_shape, const TShape& kernel_shape,
+  const TShape& pad, const TShape& stride, const TShape& dilation,
+  const uint32_t deformable_group, DType* data_col) {
+  if (2 == kernel_shape.ndim()) {
+    LOG(FATAL) << "not implemented";
```
Comment (piiswrong) [inline src/operator/contrib/nn/deformable_im2col.h]: say "only implemented in GPU"

**Comment by Orpine** [comment]
Updated, please have a check. @piiswrong

**Comment by piiswrong** [comment]
merged. Thanks.

**Comment by piiswrong** [comment]
Are you planning to submit the algorithms that use these ops as examples?

**Comment by piiswrong** [comment]
Alternatively you can put a link to your repos in example/readme

**Comment by Orpine** [comment]
@piiswrong Thanks for your advice, I've added a link #6754
